### ISSUE TITLE
How to improve inference performance when running on CPU?

### ISSUE THREAD

**Issue by zhijianli** [facebookresearch/fairseq#4490]
GPUs are too expensive, so I can only perform inference with CPUs

I try to finetune the  XLS-R 300M model with some data and then when I try to perform inference on CPU.

but it's too slow，RTF exceeds 1.
RTF:1.49
RTF:1.51
RTF:1.53
RTF:1.54
RTF:1.57
RTF:1.35
RTF:1.18
RTF:1.19
RTF:1.22
RTF:1.22
RTF:1.46
RTF:1.65
RTF:1.67
RTF:1.7
RTF:1.74
RTF:1.75
RTF:1.56
RTF:1.58
RTF:1.6
RTF:1.65
RTF:1.59
RTF:1.39

How to improve inference performance when running on CPU?

**my parameters:**
 - --w2l-decoder viterbi 
 - --lm-weight 2 
 - --word-score -10 
 - --sil-weight -5 
 - --criterion ctc 
 - --labels ltr 
 - --max-tokens 8000000 
 - --post-process letter 

I tried to adjust the parameters, Change viterbi to kenlm or fairseqlm，but the effect was not good.

I try to finetune the wav2vec small model (wav2vec_small.pt) and then perform inference，The effect is a little better, but not too much.

**environment:**
 - cpu ：40  Intel(R) Xeon(R) Gold 6138 CPU @ 2.00GHz
 - cpu cores： 20
 - fairseq Version：1.0.0a0+0dfd6b6
 - PyTorch Version ：1.10.0+cu113
 - OS ：Ubuntu 20.04.4 LTS
 - How you installed fairseq: pip3
 - Python version：3.8.5

**Comment by gmryu** [comment]
Hi, I am merely beginner in efficiency.

Did you use like `--num-workers 20` to open subprocess for loading data?
Well, I do not know how cpu handles models. I guess it is still one cpu for a model.
In that case, you may also want to run multiple generate.py

Also, I believe your `--max-tokens` is too high?
Using too much memory slows down programs.

Would you mind telling me what is RTF by the way?
